superset-frontend/src/explore/actions/hydrateExplore.ts

```diff
@@ -0,0 +1,125 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { ControlStateMapping } from '@superset-ui/chart-controls';
+
+import {
+  ChartState,
+  ExplorePageInitialData,
+  ExplorePageState,
+} from 'src/explore/types';
+import { getChartKey } from 'src/explore/exploreUtils';
+import { getControlsState } from 'src/explore/store';
+import { Dispatch } from 'redux';
+import { ensureIsArray } from '@superset-ui/core';
+import {
+  getFormDataFromControls,
+  applyMapStateToPropsToControl,
+} from 'src/explore/controlUtils';
+import { getDatasourceUid } from 'src/utils/getDatasourceUid';
+import { getUrlParam } from 'src/utils/urlUtils';
+import { URL_PARAMS } from 'src/constants';
+import { findPermission } from 'src/utils/findPermission';
+
+export const HYDRATE_EXPLORE = 'HYDRATE_EXPLORE';
+export const hydrateExplore =
+  ({ form_data, slice, dataset: datasource }: ExplorePageInitialData) =>
+  (dispatch: Dispatch, getState: () => ExplorePageState) => {
+    const initialFormData = form_data;
```

Comment:
I think we should construct an appropriate/default formData here so that supports to open dataset.

```
    const initialFormData = slice?.form_data ? slice.form_data : form_data;
    const initialData = { form_data, slice, datasource };
    const initialControls = getControlsState(
      initialData,
      initialFormData,
    ) as ControlStateMapping;
    const { user, datasources, charts } = getState();

    const exploreState = {
....
....
      controls: initialControls,
....
    }
....
....
    const sliceFormData = slice
      ? getFormDataFromControls(initialControls)
      : null;
```